### PR TITLE
Automatically forget learnt DNS server information when network config changes

### PR DESCRIPTION
**Pull request by poettering**
Also, add mechanisms to request this manually.

### CODE REVIEW

File: src/resolve/resolved-dns-server.c [hunk 1]
```diff
@@ -828,6 +828,37 @@ void dns_server_flush_cache(DnsServer *s) {
...
+        s->max_rtt = 0;
+        s->resend_timeout = DNS_TIMEOUT_MIN_USEC;
+
+        s->verified_feature_level = _DNS_SERVER_FEATURE_LEVEL_INVALID;
+        s->possible_feature_level = DNS_SERVER_FEATURE_LEVEL_BEST;
+
+        s->received_udp_packet_max = DNS_PACKET_UNICAST_SIZE_MAX;
+
+        s->n_failed_udp = 0;
+        s->n_failed_tcp = 0;
```
Comment (keszybz) [inline src/resolve/resolved-dns-server.c]: Why not call `dns_server_reset_counters()` here?

I also think `dns_server_reset` would be more in line with the other function names. This function doesn't "examine" anything and I don't think we say "reexamine" anywhere else in the systemd codebase.

File: src/resolve/resolved-manager.c [hunk 2]
```diff
@@ -1396,6 +1396,19 @@ void manager_flush_caches(Manager *m) {
...
+        Iterator i;
+        Link *l;
+
+        dns_server_reexamine_all(m->dns_servers);
+        dns_server_reexamine_all(m->fallback_dns_servers);
+
+        HASHMAP_FOREACH(l, m->links, i)
+                dns_server_reexamine_all(l->dns_servers);
+
+        log_info("Re-examining all servers.");
```
Comment (keszybz) [inline src/resolve/resolved-manager.c]: Yeah, this is misleading. "Re-examining" implies that there's some queries being sent out, while this just resets some stats.

File: man/systemd-resolve.xml [hunk 3]
```diff
@@ -299,7 +299,17 @@
...
-        <listitem><para>Flushes all DNS resource record caches the service maintains locally.</para></listitem>
+        <listitem><para>Flushes all DNS resource record caches the service maintains locally. This is mostly equivalent
+        to sending the <constant>SIGUSR2</constant> to the <command>systemd-resolved</command>
+        service.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--reexamine-servers</option></term>
+
+        <listitem><para>Flushes all DNS resource record caches the service maintains locally. This is mostly equivalent
```
Comment (keszybz) [inline man/systemd-resolve.xml]: Copy-pasto here.

File: man/systemd-resolved.service.xml [hunk 4]
```diff
@@ -202,19 +202,37 @@
...
+
+      <varlistentry>
+        <term><constant>SIGRTMIN+1</constant></term>
+
+        <listitem><para>Upon reception of the <constant>SIGRTMIN+1</constant> process signal
+        <command>systemd-resolved</command> will forget everything it learnt about the configured DNS
+        servers. Specifically any information about server feature support is flushed out, and the server feature
+        probing logic is restarted on the next request, starting with the most fully featured level. Note that it
+        should normally not be necessary to request this explicitly – except for debugging purposes – as
+        <command>systemd-resolved</command> automatically forgets this learnt information any time the DNS server
```
Comment (keszybz) [inline man/systemd-resolved.service.xml]: s/this learnt information/this information/

File: src/resolve/resolve-tool.c [hunk 5]
```diff
@@ -1588,6 +1607,7 @@ static void help(void) {
                "     --reset-statistics     Reset resolver statistics\n"
                "     --status               Show link and server status\n"
                "     --flush-caches         Flush all local DNS caches\n"
+               "     --reexamine-servers    Forget DNS server feature level\n"
```
Comment (keszybz) [inline src/resolve/resolve-tool.c]: level*s*

File: src/resolve/resolved-manager.c [hunk 6]
```diff
@@ -616,6 +627,7 @@ int manager_new(Manager **ret) {
 
         (void) sd_event_add_signal(m->event, &m->sigusr1_event_source, SIGUSR1, manager_sigusr1, m);
         (void) sd_event_add_signal(m->event, &m->sigusr2_event_source, SIGUSR2, manager_sigusr2, m);
+        (void) sd_event_add_signal(m->event, &m->sigusr2_event_source, SIGRTMIN+1, manager_sigrtmin1, m);
```
Comment (keszybz) [inline src/resolve/resolved-manager.c]: This looks wrong.

**Comment by poettering** [comment]
I force pushed a new version now, adressing all points raised. I also added another commit on top, adding some more debug output. PTAL

**Pull request merged**